### PR TITLE
[Merged by Bors] - feat: interior and boundary of a manifold

### PR DESCRIPTION
**Pull request by grunweg**
We use the standard definition, with respect to the preferred charts at each point.

Open-ness of the interior is non-trivial, hence left to a future PR:
- for instance, in finite dimensions this requires e.g. knowing the homology of spheres, which mathlib doesn't yet have.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by winstonyin** [comment]
Thanks for writing this PR! All seems quite standard; looks good to me!

**Comment by grunweg** [comment]
TL;DR. This definition seems to be forward compatible with Banach manifolds with boundary.

For the record: there's an established theory of Banach manifolds with boundary and corners also; the openness of the interior and disjointness of interior and boundary still hold in this context. (See e.g. https://arxiv.org/pdf/1910.14580.pdf, Theorem 4.12 (which references the (pay-walled) textbook *Differential topology* by Roig and Domínguez) for a reference.) Their definition is more specific than mathlib's; they consider a model space `E` given as the intersection of finitely many half-spaces.

Their proof is different from the standard finite-dimensional proof that interior and boundary are independent of the chart (using homology of spheres/Brouwer fixed point theorem).

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+-/
+
+open Set
+
+-- Let `M` be a manifold with corners over the pair `(E, H)`.
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  {H : Type*} [TopologicalSpace H] {I : ModelWithCorners 𝕜 E H}
+  {M : Type*} [TopologicalSpace M] [ChartedSpace H M] [HasGroupoid M (contDiffGroupoid 0 I)]
```
Comment (sgouezel) [inline Mathlib/Geometry/Manifold/InteriorBoundary.lean:47]:
```suggestion
  {M : Type*} [TopologicalSpace M] [ChartedSpace H M]
```
This plays no role in the file.

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+/-- `p ∈ M` is an interior point of a manifold `M` iff
+for `φ` being the preferred chart at `x`, `φ x` is an interior point of `φ.target`. -/
+def ModelWithCorners.IsInteriorPoint (x : M) := extChartAt I x x ∈ interior (range I)
```
Comment (sgouezel) [inline Mathlib/Geometry/Manifold/InteriorBoundary.lean:51]: `I` should be explicit in this definition: otherwise, when you call `IsInteriorPoint x`, there is no way Lean can guess which model space `H` you are using.

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+/-- `p ∈ M` is a boundary point of a manifold `M` iff its image in the extended chart
+lies on the boundary of the model space. -/
+def ModelWithCorners.IsBoundaryPoint (x : M) := extChartAt I x x ∈ frontier (range I)
+
+namespace SmoothManifoldWithCorners
+-- TODO: can I enable dot notation as in, say, `M.interior I`?
```
Comment (sgouezel) [inline Mathlib/Geometry/Manifold/InteriorBoundary.lean:58]: no. But if you put this in the `ModelWithCorners` namespace, you can use `I.interior M`, which is probably a good idea. I think essentially all the file should be in the `ModelWithCorners` namespace, for this reason.

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+/-- Every point is either an interior or a boundary point. -/
+lemma isInteriorPoint_or_isBoundaryPoint (x : M) : I.IsInteriorPoint x ∨ I.IsBoundaryPoint x := by
+  by_cases h : extChartAt I x x ∈ interior (range I)
+  · exact Or.inl h
+  · right -- Otherwise, we have a boundary point.
+    rw [I.isBoundaryPoint_iff, ← closure_diff_interior, I.closed_range.closure_eq]
+    exact ⟨mem_range_self _, h⟩
+
+/-- A manifold decomposes into interior and boundary. -/
+lemma univ_eq_interior_union_boundary : (SmoothManifoldWithCorners.interior I M) ∪
```
Comment (sgouezel) [inline Mathlib/Geometry/Manifold/InteriorBoundary.lean:85]:
```suggestion
lemma interior_union_boundary_eq_univ : (SmoothManifoldWithCorners.interior I M) ∪
```

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    (SmoothManifoldWithCorners.boundary I M) = (univ : Set M) :=
+  le_antisymm (fun _ _ ↦ trivial) (fun x _ ↦ isInteriorPoint_or_isBoundaryPoint x)
+
+/-- The interior and boundary of a manifold `M` are disjoint. -/
+lemma interior_boundary_disjoint :
+    (SmoothManifoldWithCorners.interior I M) ∩ (SmoothManifoldWithCorners.boundary I M) = ∅ := by
```
Comment (sgouezel) [inline Mathlib/Geometry/Manifold/InteriorBoundary.lean:91]:
```suggestion
lemma disjoint_interior_boundary :
    Disjoint (SmoothManifoldWithCorners.interior I M) (SmoothManifoldWithCorners.boundary I M) := by
```
That's the standard spelling of disjointess.

File: Mathlib/Topology/Basic.lean [hunk 6]
```diff
@@ -709,6 +709,11 @@ theorem closure_diff_interior (s : Set α) : closure s \ interior s = frontier s
   rfl
 #align closure_diff_interior closure_diff_interior
 
+/-- Interior and frontier are disjoint. -/
+lemma interior_frontier_disjoint : interior s ∩ frontier s = ∅ := by
```
Comment (sgouezel) [inline Mathlib/Topology/Basic.lean:713]:
```suggestion
lemma disjoint_interior_frontier : Disjoint (interior s) (frontier s) := by
```

**Review by sgouezel — COMMENTED** [review]
LGTM, thanks!

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Geometry/Manifold/InteriorBoundary.lean]: That sounds almost as good - I've adopted this now.

**Review by sgouezel — COMMENTED** [review]
bors r+
Thanks!

**Comment by grunweg** [comment]
Thank you for the review!
